File: python/deepstrike/safety/permissions.py
from __future__ import annotations
from dataclasses import dataclass
from enum import Enum


class PermissionMode(Enum):
    DEFAULT = "default"
    PLAN = "plan"
    AUTO = "auto"


@dataclass
class Permission:
    tool: str
    action: str
    allowed: bool = True
    requires_approval: bool = False
    note: str = ""


@dataclass
class PermissionDecision:
    allowed: bool
    reason: str = ""
    requires_approval: bool = False
    matched_permission: Permission | None = None


class PermissionManager:
    def __init__(self, mode: PermissionMode = PermissionMode.DEFAULT):
        self.permissions: dict[str, Permission] = {}
        self.mode = mode

    def grant(self, tool: str, action: str, *, requires_approval: bool = False, note: str = ""):
        self.permissions[self._key(tool, action)] = Permission(
            tool=tool, action=action, allowed=True, requires_approval=requires_approval, note=note
        )

    def revoke(self, tool: str, action: str, *, note: str = ""):
        self.permissions[self._key(tool, action)] = Permission(
            tool=tool, action=action, allowed=False, note=note
        )

    def check(self, tool: str, action: str) -> tuple[bool, str]:
        decision = self.evaluate(tool, action)
        return decision.allowed, decision.reason

    def evaluate(self, tool: str, action: str) -> PermissionDecision:
        if self.mode == PermissionMode.PLAN:
            return PermissionDecision(allowed=False, reason="Plan mode: execution not allowed")

        permission = self._match_permission(tool, action)
        if permission:
            if not permission.allowed:
                return PermissionDecision(allowed=False, reason=permission.note or "Permission denied", matched_permission=permission)
            if permission.requires_approval and self.mode != PermissionMode.AUTO:
                return PermissionDecision(allowed=False, reason=permission.note or "Approval required", requires_approval=True, matched_permission=permission)
            return PermissionDecision(allowed=True, matched_permission=permission)

        if self.mode == PermissionMode.AUTO:
            return PermissionDecision(allowed=True, reason="Auto mode")

        return PermissionDecision(allowed=False, reason="No explicit permission")

    def _match_permission(self, tool: str, action: str) -> Permission | None:
        for key in [self._key(tool, action), self._key(tool, "*"), self._key("*", action), self._key("*", "*")]:
            if key in self.permissions:
                return self.permissions[key]
        return None

    def _key(self, tool: str, action: str) -> str:
        return f"{tool}:{action}"

File: node/src/safety/permissions.ts
export enum PermissionMode {
  DEFAULT = "DEFAULT",
  PLAN = "PLAN",
  AUTO = "AUTO",
}

export interface Permission {
  tool: string
  action: string
  allowed: boolean
  requiresApproval: boolean
  note: string
}

export interface PermissionDecision {
  allowed: boolean
  reason: string
  requiresApproval?: boolean
  matchedPermission?: Permission
}

export class PermissionManager {
  private permissions = new Map<string, Permission>()

  constructor(private mode: PermissionMode = PermissionMode.DEFAULT) {}

  grant(resource: string, action: string, options: { requiresApproval?: boolean; note?: string } = {}): this {
    const key = `${resource}:${action}`
    this.permissions.set(key, {
      tool: resource, action, allowed: true,
      requiresApproval: options.requiresApproval ?? false,
      note: options.note ?? "",
    })
    return this
  }

  grantWithApproval(resource: string, action: string, note = ""): this {
    return this.grant(resource, action, { requiresApproval: true, note })
  }

  revoke(resource: string, action: string, note = ""): this {
    const key = `${resource}:${action}`
    this.permissions.set(key, { tool: resource, action, allowed: false, requiresApproval: false, note })
    return this
  }

  private matchPermission(tool: string, action: string): Permission | undefined {
    for (const key of [`${tool}:${action}`, `${tool}:*`, `*:${action}`, `*:*`]) {
      const p = this.permissions.get(key)
      if (p) return p
    }
    return undefined
  }

  evaluate(resource: string, action: string): PermissionDecision {
    if (this.mode === PermissionMode.AUTO) return { allowed: true, reason: "AUTO mode" }
    if (this.mode === PermissionMode.PLAN) return { allowed: false, reason: "PLAN mode blocks all" }
    const perm = this.matchPermission(resource, action)
    if (!perm) return { allowed: false, reason: "not granted" }
    if (!perm.allowed) return { allowed: false, reason: perm.note || "permission denied", matchedPermission: perm }
    if (perm.requiresApproval) return { allowed: false, reason: perm.note || "requires approval", requiresApproval: true, matchedPermission: perm }
    return { allowed: true, reason: "granted", matchedPermission: perm }
  }
}
